frontend/src/Frontend/VersionedStore/V2.hs

```diff
@@ -0,0 +1,216 @@
+import Frontend.Crypto.Class
+
+-- WARNING: Upstream deps. Check this when we bump pact and obelisk!
+-- May be worth storing this in upstream independent datatypes.
+import Pact.Types.ChainId (ChainId(ChainId))
+import Pact.Types.ChainMeta (PublicMeta (..))
+import Obelisk.OAuth.Common (AccessToken, OAuthState)
+
+data StoreFrontend key a where
+  StoreFrontend_Wallet_Keys :: StoreFrontend key (KeyStorage key)
+  StoreFrontend_Wallet_Accounts :: StoreFrontend key AccountStorage
+
+  StoreFrontend_Network_PublicMeta :: StoreFrontend key PublicMeta
+  StoreFrontend_Network_Networks :: StoreFrontend key (Map NetworkName [NodeRef])
+  StoreFrontend_Network_SelectedNetwork :: StoreFrontend key NetworkName
+
+  StoreFrontend_OAuth_Tokens :: StoreFrontend key (Map OAuthProvider AccessToken)
+  StoreFrontend_OAuth_State :: OAuthProvider -> StoreFrontend key OAuthState
+
+  StoreFrontend_Gist_GistRequested :: StoreFrontend key (GistMeta, Text)
+
+  StoreFrontend_ModuleExplorer_SessionFile :: StoreFrontend key Text
+
+deriving instance Show (StoreFrontend key a)
+
+upgradeFromV0 :: (Monad m, HasCrypto key m) => DMap (V0.StoreFrontend key) Identity -> m (DMap (StoreFrontend key) Identity)
+upgradeFromV0 v0 = do
+  (newKeysList, newAccountStorage) <- foldMapM splitOldKey oldKeysList
+  let newKeys = IntMap.fromList newKeysList
+  pure $ DMap.fromList . catMaybes $
+    [ copyKeyDSum V0.StoreNetwork_PublicMeta StoreFrontend_Network_PublicMeta v0
+    , copyKeyDSum V0.StoreNetwork_SelectedNetwork StoreFrontend_Network_SelectedNetwork v0
+    -- Technically these are session only and shouldn't be here given the backup restore only works on
+    -- local storage, but desktop ignores the session vs local distinction so migrating them probably
+    -- does some good and certainly doesn't hurt.
+    -- Also, this is currently being very lazy not leaning on the Universe instance of OAuthProvider
+    , copyKeyDSum V0.StoreOAuth_Tokens StoreFrontend_OAuth_Tokens v0
+    , copyKeyDSum (V0.StoreOAuth_State OAuthProvider_GitHub) (StoreFrontend_OAuth_State OAuthProvider_GitHub) v0
+
+    , copyKeyDSum V0.StoreModuleExplorer_SessionFile StoreFrontend_ModuleExplorer_SessionFile v0
+
+    , Just (StoreFrontend_Wallet_Keys :=> Identity newKeys)
+    , Just (StoreFrontend_Wallet_Accounts :=> Identity newAccountStorage)
+    , newNetworks
+    ]
+  where
+    oldKeysList = maybe [] (IntMap.toList . runIdentity) (DMap.lookup V0.StoreWallet_Keys v0)
+
+    -- We have to walk through the slightly different encoding of the Network information.
+    -- Also if the storage contains _no_ network configuration then we shouldn't break the new version
+    -- by storing an empty object.
+    newNetworks = (\nets -> StoreFrontend_Network_Networks :=> Identity (convertNodeRefs $ V0.unNetworkMap $ runIdentity nets))
+      <$> DMap.lookup V0.StoreNetwork_Networks v0
+
+    -- This will regenerate the missing key. Desktop will recover the key with
```

Comment:
This is something we really need to look out for with the upcoming chainweaver changes